packages/plugins/rpc-manager/src/lib/index.ts

```diff
@@ -20,25 +22,78 @@ export default class RpcManager {
   private events: EmbarkEvents;
   public _accounts: any[] | null = null;
   public _nodeAccounts: any[] | null = null;
+  private blockchainConfig: EmbarkConfig["blockchainConfig"];
   constructor(private readonly embark: Embark) {
     this.events = embark.events;
     this.logger = embark.logger;
     this.rpcModifierEvents = new Events() as EmbarkEvents;
+    this.blockchainConfig = this.embark.config.blockchainConfig;
     this.init();
   }
 
-  private async init() {
+  protected get web3() {
+    return (async () => {
+      if (!this._web3) {
+        let provider;
+        let testsCliOptions;
+        let isTest = false;
 
-    this.rpcModifierEvents.setCommandHandler("nodeAccounts:updated", this.updateAccounts.bind(this));
-    this.rpcModifierEvents.setCommandHandler("nodeAccounts:added", async (addedNodeAccount: any, cb: Callback<null>) => {
+        await this.events.request2("blockchain:started");
+
+        if (this.embark.currentContext.includes(constants.contexts.test)) {
+          isTest = true;
+          testsCliOptions = await this.events.request2("tests:cli:options");
+        }
+        if (isTest && testsCliOptions.node === constants.blockchain.vm) {
+          provider = await this.events.request2("blockchain:client:vmProvider");
+        } else {
+          let endpoint = this.blockchainConfig.endpoint;
+          if (isTest && testsCliOptions.node) {
+            endpoint = testsCliOptions.node;
+          }
+          provider = await this.events.request2("blockchain:client:provider", "ethereum", endpoint);
+        }
+        this._web3 = new Web3(provider);
+      }
+      return this._web3;
+    })();
+  }
+
+  private get nodeAccounts() {
+    return (async () => {
       if (!this._nodeAccounts) {
-        this._nodeAccounts = [addedNodeAccount];
-      } else {
-        this._nodeAccounts.push(addedNodeAccount);
+        const web3 = await this.web3;
+        this._nodeAccounts = await web3.eth.getAccounts();
+      }
+      return this._nodeAccounts || [];
+    })();
+  }
+
+  private get accounts() {
+    return (async () => {
+      if (!this._accounts) {
+        const web3 = await this.web3;
+        const nodeAccounts = await this.nodeAccounts;
+        this._accounts = AccountParser.parseAccountsConfig(this.embark.config.blockchainConfig.accounts, web3, dappPath(), this.logger, nodeAccounts);
       }
-      return this.updateAccounts(this._nodeAccounts, cb);
+      return this._accounts || [];
+    })();
+  }
+
+  private async init() {
+
+    this.embark.registerActionForEvent("tests:config:updated", { priority: 40 }, (_params, cb) => {
+      // web.eth.getAccounts may return a different value now
+      // update accounts across all modifiers
+      this.updateAccounts(cb);
```

Comment:
Here, we should also reset `_web3`, because the blockchain configs could have changed.

Reply:
Updated, good catch!